server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java

```diff
@@ -78,31 +82,51 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
     protected void shardOperationOnPrimary(
         BasicReplicationRequest shardRequest,
         IndexShard primary,
-        ActionListener<PrimaryResult<ReplicaShardRefreshRequest, ReplicationResponse>> listener
+        ActionListener<PrimaryResult<BasicReplicationRequest, ReplicationResponse>> listener
     ) {
-        ActionListener.completeWith(listener, () -> {
+        try (
+            var listeners = new RefCountingListener(
+                listener.map(ignored -> { return new PrimaryResult<>(shardRequest, new ReplicationResponse()); })
+            )
+        ) {
             var refreshResult = primary.refresh(SOURCE_API);
             logger.trace("{} refresh request executed on primary", primary.shardId());
-            var shardRefreshRequest = new ReplicaShardRefreshRequest(
-                primary.shardId(),
-                shardRequest.getParentTask(),
-                refreshResult.generation()
-            );
-            return new PrimaryResult<>(shardRefreshRequest, new ReplicationResponse());
-        });
+
+            // Forward the request to all nodes that hold unpromotable replica shards
+            final ClusterState clusterState = clusterService.state();
+            clusterService.state()
+                .routingTable()
+                .shardRoutingTable(shardRequest.shardId())
+                .replicaShards()
```

Comment:
Slight preference for this:

```suggestion
                .assignedShards()
```